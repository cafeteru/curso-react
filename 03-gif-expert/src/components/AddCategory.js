import { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ({ setCategories }) => {
  const defaultValue = "";
  const [inputValue, setInputValue] = useState(defaultValue);
  const handleInputChange = (e) => setInputValue(e.target.value);
  const handleSubmit = (event) => {
    event.preventDefault(); // Evita recarga de la pagina cuando envia el formulario
    if (inputValue.trim().length > 2) {
      setCategories((categories) => [inputValue, ...categories]);
      setInputValue(defaultValue);
    }
  };
  return (
    <form onSubmit={handleSubmit}>
      <input type="text" value={inputValue} onChange={handleInputChange} />
    </form>
  );
};

AddCategory.propTypes = { setCategories: PropTypes.func.isRequired };
