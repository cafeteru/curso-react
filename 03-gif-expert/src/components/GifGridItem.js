export const GifGridItem = ({ title, url }) => {
  return (
    <div className={"card"}>
      <img alt={title} src={url} />
      <p>{title}</p>
    </div>
  );
};
