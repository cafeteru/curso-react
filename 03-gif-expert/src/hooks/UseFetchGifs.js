import { useEffect, useState } from "react";
import { getGifs } from "../helpers/GetGifs";

export const useFetchGifs = (category) => {
  const [state, setState] = useState({
    gifs: [],
    loading: true,
  });

  // useEffect -> para controlar las llamadas
  useEffect(() => {
    getGifs(category).then((res) => setState({ gifs: res, loading: false }));
  }, [category]);

  return state;
};
