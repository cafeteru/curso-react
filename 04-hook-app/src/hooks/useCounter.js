import { useState } from "react";

export const useCounter = (initialState = 10) => {
  const [count, setCount] = useState(initialState);

  const increment = (value = 1) => setCount(count + value);
  const decrement = (value = 1) => setCount(count - value);
  const reset = () => setCount(0);

  return {
    count,
    increment,
    decrement,
    reset,
  };
};
