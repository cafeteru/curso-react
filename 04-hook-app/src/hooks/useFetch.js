import { useEffect, useState } from "react";

export const useFetch = (url) => {
  let initialState = {
    data: null,
    loading: true,
    error: null,
  };
  const [state, setState] = useState(initialState);

  useEffect(() => {
    setState({
      ...initialState,
      loading: true,
    });
    fetch(url)
      .then((resp) => resp.json())
      .then((data) =>
        setState({
          loading: false,
          error: null,
          data,
        })
      );
  }, [url]);

  return state;
};
