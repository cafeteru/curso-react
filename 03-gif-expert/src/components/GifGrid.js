import { useFetchGifs } from "../hooks/UseFetchGifs";
import { GifGridItem } from "./GifGridItem";
import PropTypes from "prop-types";

export const GifGrid = ({ category }) => {
  const { gifs, loading } = useFetchGifs(category);
  return (
    <>
      <h3>{category}</h3>
      {loading && <p>"Cargando..."</p>}
      <div className={"card-grid"}>
        {gifs.map((img) => (
          <GifGridItem key={img.id} {...img} />
        ))}
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired,
};
