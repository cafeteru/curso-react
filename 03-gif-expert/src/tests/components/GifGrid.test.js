import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/UseFetchGifs";

jest.mock("../../hooks/UseFetchGifs"); // crea el mock

describe("GifGrid", () => {
  const category = "Bleach";
  test("Debe de mostrar crearse", () => {
    useFetchGifs.mockReturnValue({
      gifs: [],
      loading: true,
    });

    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe de mostrar items cuando se cargan imagenes useFetchGifs", () => {
    const gifs = [
      {
        id: "abs",
        url: "http://localhost",
        title: "title",
      },
    ];
    useFetchGifs.mockReturnValue({
      gifs,
      loading: true,
    });
    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper.find("p").exists()).toBeFalsy();
  });
});
