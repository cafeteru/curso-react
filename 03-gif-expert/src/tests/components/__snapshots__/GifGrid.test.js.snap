// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifGrid Debe de mostrar crearse 1`] = `
<Fragment>
  <h3>
    Bleach
  </h3>
  <p>
    "Cargando..."
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
